### DISCUSSION TITLE
Utils module or crate?

### DISCUSSION THREAD

**Discussion by TheAwiteb** [ratatui/ratatui#1173] [Discussion]
When I discover the Ratatui book, I found [this amazing util] written by @Nickiel12 I think, the util centered a box in the middle of an area. And I wondered, why is there no module or crate that contains such awesome utils (I'm sure there are more that I don't know about).

### The util

https://github.com/ratatui-org/ratatui-website/blob/f9f8ceb01f6a9e9ae5f93cd5912126a991976876/code/ratatui-json-editor-app/src/ui.rs#L183-L205

[this amazing util]: https://ratatui.rs/tutorials/json-editor/closing-thoughts/#uirs

**Comment by orhun** [comment]
AFAIK there are a couple of other things that are similar to this:

- https://github.com/joshka/ratatui-widgets
- https://github.com/ratatui-org/ratatui-macros
- https://github.com/sophacles/extra-widgets (outdated)

Maybe we can increase the visibility of repos like this in the documentation.

**Reply by joshka** [reply]
There's a couple of more specific crates that meet the need:
- https://crates.io/crates/tui-popup (wraps pretty much any widget in a popup that is centered)
- https://crates.io/crates/tui_confirm_dialog (presents a dialog to confirm and is interactive)

Also, since that tutorial was written, we introduced Flex alignment and Fill. There's two better ways to express this depending on whether you need the side areas or not:
- Flex https://docs.rs/ratatui/latest/ratatui/layout/enum.Flex.html
```rust
let [center] = Layout::horizontal([Constraint::Percentage(x_percentage)]).flex(Flex::Center).areas(area);
let [center] = Layout::vertical([Constraint::Percentage(y_percentage)]).flex(Flex::Center).areas(area);
```
- Constraint::Fill https://docs.rs/ratatui/latest/ratatui/layout/enum.Constraint.html#variant.Fill
```rust
let [left, center, right] = Layout::horizontal([
    Constraint::Fill(1),
    Constraint::Percentage(x_percentage),
    Constraint::Fill(1)])
.flex(Flex::Center).areas(area);
let [top, center, bottom] = Layout::vertical([
    Constraint::Fill(1),
    Constraint::Percentage(y_percentage),
    Constraint::Fill(1)])
.flex(Flex::Center).areas(center);
```

P.s. Ratatui-widgets is very much in its infancy and underdeveloped.
